subs/pantry/src/Pantry/Storage.hs

```diff
@@ -436,30 +449,122 @@ loadHackageTarballInfo name version = do
+    => SafeFilePath
+    -> ReaderT SqlBackend (RIO env) (Entity FilePath)
+loadFilePath path = do
+    fp <- getBy $ UniqueSfp path
+    case fp of
+        Nothing ->
+            error $
+            "loadFilePath: No row found for " <>
+            (T.unpack $ P.unSafeFilePath path)
+        Just record -> return record
+
+generateHPack ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => TreeId
+    -> VersionId
+    -> ReaderT SqlBackend (RIO env) (Key HPack)
+generateHPack tid vid = do
+    let hpackPath =
+            maybe
+                (error
+                     "generateHPack: Not able to convert hpack file to SafeFilePath")
+                id $
+            P.mkSafeFilePath (T.pack Hpack.packageConfig)
+    filepath <- loadFilePath hpackPath
+    let filePathId :: FilePathId = entityKey filepath
+    hpackTreeEntry <-
+        selectFirst [TreeEntryTree ==. tid, TreeEntryPath ==. filePathId] []
+    hpackEntity <-
+        case hpackTreeEntry of
+            Nothing ->
+                error $
+                "generateHPack: No package.yaml file found in TreeEntry for TreeId:  " ++
+                (show tid)
+            Just record -> return record
+    hpackBS <- loadBlobById (treeEntryBlob $ entityVal hpackEntity)
+    (pkgName, cabalBS) <- lift $ hpackToCabal hpackBS
+    bid <- storeCabalFile cabalBS pkgName
+    let cabalFile = P.cabalFileName pkgName
+    fid <- insertBy FilePath {filePathPath = cabalFile}
+    let hpackRecord =
+            HPack
+                { hPackTree = tid
+                , hPackVersion = vid
+                , hPackCabal = bid
+                , hPackPath = getKey fid
+                }
+    val <- insertBy hpackRecord
+    return $ getKey val
+
+
+hpackVersionId ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => ReaderT SqlBackend (RIO env) VersionId
+hpackVersionId = do
+    hpackSoftwareVersion <- lift $ hpackVersion
+    hid <- insertBy $ Version {versionVersion = P.VersionP hpackSoftwareVersion}
+    return $ getKey hid
+
+storeHPack ::
+       (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => TreeId
+    -> ReaderT SqlBackend (RIO env) (Key HPack)
+storeHPack tid = do
+    vid <- hpackVersionId
+    hpackRecord <- getBy (UniqueHPack tid vid)
+    case hpackRecord of
+      Nothing -> generateHPack tid vid
+      Just record -> return $ entityKey record
+
 storeTree
-  :: (HasPantryConfig env, HasLogFunc env)
+  :: (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
   => P.PackageIdentifier
   -> P.Tree
   -> P.TreeEntry
   -- ^ cabal file
+  -> P.BuildFileType
   -> ReaderT SqlBackend (RIO env) (TreeId, P.TreeKey)
-storeTree (P.PackageIdentifier name version) tree@(P.TreeMap m) (P.TreeEntry (P.BlobKey cabal _) cabalType) = do
+storeTree (P.PackageIdentifier name version) tree@(P.TreeMap m) (P.TreeEntry (P.BlobKey btypeSha _) fileType) btype = do
```

Comment:
did you want to use `buildFileType` instead of `btype`?

Reply:
Done.